javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/InstrumentationModuleInstaller.java

```diff
@@ -69,44 +72,87 @@ AgentBuilder install(
     }
 
     if (instrumentationModule.isIndyModule()) {
-      return installIndyModule(instrumentationModule, parentAgentBuilder);
+      return installIndyModule(instrumentationModule, parentAgentBuilder, config);
     } else {
       return installInjectingModule(instrumentationModule, parentAgentBuilder, config);
     }
   }
 
   private AgentBuilder installIndyModule(
-      InstrumentationModule instrumentationModule, AgentBuilder parentAgentBuilder) {
-
-    IndyModuleRegistry.registerIndyModule(instrumentationModule);
-
+      InstrumentationModule instrumentationModule,
+      AgentBuilder parentAgentBuilder,
+      ConfigProperties config) {
+    List<String> helperClassNames =
+        InstrumentationModuleMuzzle.getHelperClassNames(instrumentationModule);
     HelperResourceBuilderImpl helperResourceBuilder = new HelperResourceBuilderImpl();
     instrumentationModule.registerHelperResources(helperResourceBuilder);
+    List<TypeInstrumentation> typeInstrumentations = instrumentationModule.typeInstrumentations();
+    if (typeInstrumentations.isEmpty()) {
+      if (!helperClassNames.isEmpty() || !helperResourceBuilder.getResources().isEmpty()) {
+        logger.log(
+            WARNING,
+            "Helper classes and resources won't be injected if no types are instrumented: {0}",
+            instrumentationModule.instrumentationName());
+      }
+
+      return parentAgentBuilder;
+    }
+
+    List<String> injectedHelperClassNames = Collections.emptyList();
+    if (instrumentationModule instanceof ExperimentalInstrumentationModule) {
+      ExperimentalInstrumentationModule experimentalInstrumentationModule =
+          (ExperimentalInstrumentationModule) instrumentationModule;
+      injectedHelperClassNames = experimentalInstrumentationModule.injectedClassNames();
+      helperClassNames.removeAll(injectedHelperClassNames);
+    }
+
+    IndyModuleRegistry.registerIndyModule(instrumentationModule);
 
     ClassInjectorImpl injectedClassesCollector = new ClassInjectorImpl(instrumentationModule);
     if (instrumentationModule instanceof ExperimentalInstrumentationModule) {
       ((ExperimentalInstrumentationModule) instrumentationModule)
           .injectClasses(injectedClassesCollector);
     }
 
+    // TODO (Jonas): Adapt MuzzleMatcher to use the same type lookup strategy as the
+    // InstrumentationModuleClassLoader (see IndyModuleTypePool)
+    MuzzleMatcher muzzleMatcher =
+        new MuzzleMatcher(logger, instrumentationModule, config) {
+          @Override
+          public boolean matches(
+              TypeDescription typeDescription,
+              ClassLoader classLoader,
+              JavaModule module,
+              Class<?> classBeingRedefined,
+              ProtectionDomain protectionDomain) {
+            ClassLoader instrumentationClassLoader =
+                IndyModuleRegistry.getInstrumentationClassloader(
+                    instrumentationModule.getClass().getName(), classLoader);
```

Comment:
Nit: the `MuzzleMatcher` already gets the instrumentation module in the constructor, it can check `isIndyModule()` itself and override the CL by itself.

Reply:
done.